### ISSUE TITLE
build error in debball DeviceManagerVfr.obj in .../ovmf-dir-remote/BaseTools/Source/C fatal BootManagerVfr.c no such file

### ISSUE THREAD

**Issue by cvcachagua** [tklengyel/drakvuf#1483]
Not sure what went wrong

I performed in the ...drakvuf/xen directory: $ make -j4 debball |tee log -- and from the log:
and in .../drakvuf/xen/tools/firmware/ovmf-dir-remote/BaseTools/Source/C
Error 1 in .../FileExplorerLib/OUTPUT/FileExplorerVfr.1 BaseTools C Tool binary was not found (VfrCompile)
cc1: fatal error: .../DEBUG/BootManagerVfr.c: no such file or directory
Ubuntu Server 20.04 

Previously in .../drakvuf/ directory $ sudo apt install libsdl-dev -- did not install, so I installed libsdl1.2-dev and libsdl1.2-compat-dev

also python-dev did not install, so I installed python3-pip and python3-dev -- trying to follow directions from the web site drakvuf.com

Please help.

**Comment by tklengyel** [comment]
If you don't need ovmf then just configure xen without the --enable-ovmf flag.

**Comment by cvcachagua** [comment]
my bad the previous step failed as I tried a completely fresh install -- I should have sent standard err to output too.
make -j4 dist-tools -- was the problem.

I will have to look up ovmf, not sure what it does, but from what you say, I probably do not need it.  I am primarily concerned with process injection.

**Comment by cvcachagua** [comment]
I was developing on my laptop at home, so nice to have feature, but I can work around this by going to work -- sad.
Any chance it can be fixed?

**Comment by cvcachagua** [comment]
I assume in config you remove the --enable-ovmf flag -- I tried that but still got an issue: some shell if statement describing that some file relating to ovmf did not exist.  

Is this coming from xen, or drakvuf?

Any idea when this feature last worked?  Perhaps I can just roll back to the last working version. 

I appreciate the help Tamas.

**Comment by tklengyel** [comment]
It's coming from Xen. Afaict it still works on Ubuntu 20.04, if you are using something newer you might run into issues.

**Comment by cvcachagua** [comment]
you were correct, I used 22Jelly. I switched to 20.04.4 -- and it built.  I noticed I had kernel 5.4.0-122-generic.
I see available 5.15.0-41.44 to install (apt-cache search |grep 5.15.0-41.44)
Post install, I assume that I will need to build all over again -- Is this correct?  (if so, I wish the kernel check step was before the build in drakvuf.com)

**Comment by tklengyel** [comment]
No, if you just install a newer kernel everything else should just keep working.

**Comment by cvcachagua** [comment]
great news -- it took 3hrs to build dist-tools. Thanks

**Comment by cvcachagua** [comment]
I am getting a message in /var/log/dmesg
[    3.905696] systemd[1]: proc-xen.mount: Failed to check directory /proc/xen: No such file or directory
[    3.907148] systemd[1]: Mounting /proc/xen...
[    3.977074] systemd[1]: proc-xen.mount: Mount process exited, code=exited, status=32/n/a
[    3.979087] systemd[1]: proc-xen.mount: Failed with result 'exit-code'.
[    3.979838] systemd[1]: Failed to mount /proc/xen.

yet my fstab seems ok (one of the boots is a windows partition, not currently used):
/dev/disk/by-id/dm-uuid-LVM-jxTl7KP7p1LGZtHpIdCBk6nWS16uzcQQo2XysVeOxiFCXISPv3LpMZAGIsu8y4sQ / ext4 defaults 0 1
# /boot was on /dev/nvme0n1p2 during curtin installation
/dev/disk/by-uuid/ecdd0caa-ddf3-4b5e-a5c6-8273a7880826 /boot ext4 defaults 0 1
# /boot/efi was on /dev/nvme0n1p1 during curtin installation
/dev/disk/by-uuid/644E-D828 /boot/efi vfat defaults 0 1
/swap.img	none	swap	sw	0	0
none /proc/xen xenfs defaults,nofail 0 0

Any ideas what may be wrong, I have the executables xen-*?

**Comment by tklengyel** [comment]
Did you boot into Xen from grub?

**Comment by cvcachagua** [comment]
as far as I can tell I did.  In grub I have:
GRUB_CMDLINE_XEN_DEFAULT="dom0_mem=4096M,max:4096M dom0_max_vcpus=4 dom0_vcpus_pin=1 force-ept=1 ept=ad=0 hap_1gb=0 hap_2mb=0 altp2m=1 hpet=legacy-replacement smt=0"

I used -- update-grub

In syslog (a couple of things I noticed in correct timing order): 
kernel: [     0.698913]     BOOT_IMAGE=/vmlinuz-5.15.0-41-generic
systemd[1]: Starting LSB: Start/stop xenstored and xenconsoled...
systemd[1]: Started LSB: Start/stop xenstored and xenconsoled.
systemd[1]: grub-common.service: Succeeded.
systemd[1]: Finished Record successful boot for GRUB.
mount[440]: mount: /proc/xen: mount point does not exist.

..... but I am getting confused between boots so I logrotate -f and reboot ....
Now I dont get xen at all and I see cloud-init

looking a bit further I check for cloud packages:
$ dpkg --get-selections |grep cloud
cloud-guest-utils
cloud-init
cloud-initramfs-copymods
cloud-initramfs-dyn-netconf

.... so I did a $ sudo apt purge of these packages, and then just to make sure I performed a $ sudo touch /etc/cloud/cloud-init.disabled, checked systemctl -a and all modules are loaded, the mount failed.  I performed a $ sudo update-grub just in case, then Performed a shutdown and then restart...

..... and again no xen-detect, looking through the /var/log/syslog I see again the mount /proc/xen mount point does not exist, and systemd-modules-load[442]: Failed to insert module 'xen_evtchn': no such device....etc.

..... what should I look for?

**Comment by tklengyel** [comment]
When you boot you also have to select the Xen entry in the grub menu. Or change the default order grub creates from /etc/grub.d

**Comment by cvcachagua** [comment]
I updated /etc/default/grub commenting out "GRUB_TIMEOUT_STYLE=hidden" and "GRUB_TIMEOUT=0"
I then reboot and choose from the list:
"Ubuntu GNU/Linux, with the Xen hypervisor" .... I had to change my secure boot setting to "disable", and it worked.

Thanks!!

... any pointers on "process injection"?

**Comment by cvcachagua** [comment]
I noticed that vgscan shows two identically named VGs:
  Cache: Duplicate VG name ubuntu-vg: Existing jxTl7K-P7p1-LGZt-HpId-CBk6-nWS1-6uzcQQ (created here) takes precedence over HdDmbE-I4eD-qZM0-mIJL-CFvU-BRpf-7rsnbq
  Found volume group "ubuntu-vg" using metadata type lvm2
  Archiving volume group "ubuntu-vg" metadata (seqno 12).
  Archiving volume group "ubuntu-vg" metadata (seqno 2).
  Creating volume group backup "/etc/lvm/backup/ubuntu-vg" (seqno 2).

I am not sure if kpartx did this, or what did.
Has anybody notice this kind of problem.
vgdisplay shows:
--- Volume group ---
  VG Name               ubuntu-vg
  System ID             
  Format                lvm2
  Metadata Areas        1
  Metadata Sequence No  2
  VG Access             read/write
  VG Status             resizable
  MAX LV                0
  Cur LV                1
  Open LV               0
  Max PV                0
  Cur PV                1
  Act PV                1
  VG Size               <33.50 GiB
  PE Size               4.00 MiB
  Total PE              8575
  Alloc PE / Size       4287 / <16.75 GiB
  Free  PE / Size       4288 / 16.75 GiB
  VG UUID               HdDmbE-I4eD-qZM0-mIJL-CFvU-BRpf-7rsnbq

  --- Volume group ---
  VG Name               ubuntu-vg
  System ID             
  Format                lvm2
  Metadata Areas        1
  Metadata Sequence No  12
  VG Access             read/write
  VG Status             resizable
  MAX LV                0
  Cur LV                3
  Open LV               2
  Max PV                0
  Cur PV                1
  Act PV                1
  VG Size               <928.96 GiB
  PE Size               4.00 MiB
  Total PE              237813
  Alloc PE / Size       30720 / 120.00 GiB
  Free  PE / Size       207093 / <808.96 GiB
  VG UUID               jxTl7K-P7p1-LGZt-HpId-CBk6-nWS1-6uzcQQ

***the first one is the VM.

pvs   
  PV                           VG        Fmt  Attr PSize    PFree   
  /dev/mapper/ubuntu--vg-vm1p3 ubuntu-vg lvm2 a--   <33.50g   16.75g
  /dev/nvme0n1p3               ubuntu-vg lvm2 a--  <928.96g <808.96g

lvs
  LV        VG        Attr       LSize   Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  bkupvm1   ubuntu-vg -wi-a-----  35.00g                                                    
  ubuntu-lv ubuntu-vg -wi------- <16.75g                                                    
  ubuntu-lv ubuntu-vg -wi-ao----  50.00g                                                    
  vm1       ubuntu-vg -wi-ao----  35.00g 

Has anybody seen this problem?   I can vgrename the first VG using UUID, perhaps this is from kpartx, and then it will mount.
Appreciate any help with this, Thanks.

**Comment by cvcachagua** [comment]
created a new VM and I was able to get it to mount

dom0 $ sudo kpartx -l /dev/ubuntu-vg/vm2   --  this lists out what kpartx -a /dev/ubuntu-vg/vm2, will create in partition 
ubuntu--vg-vm2p1 : 0 2048 /dev/ubuntu-vg/vm2 2048
ubuntu--vg-vm2p2 : 0 3145728 /dev/ubuntu-vg/vm2 4096
ubuntu--vg-vm2p3 : 0 70248448 /dev/ubuntu-vg/vm2 3149824

dom0 $ sudo kpartx -a /dev/ubuntu-vg/vm2

dom0 $ lsblk
...,etc,...
nvme0n1                   259:0    0 931.5G  0 disk
├─nvme0n1p1               259:1    0   1.1G  0 part /boot/efi
├─nvme0n1p2               259:2    0   1.5G  0 part /boot
└─nvme0n1p3               259:3    0   929G  0 part
  ├─ubuntu--vg-ubuntu--lv 253:0    0    50G  0 lvm  /
  ├─ubuntu--vg-bkupvm1    253:1    0    35G  0 lvm
  ├─ubuntu--vg-vm1        253:2    0    35G  0 lvm
  └─ubuntu--vg-vm2        253:3    0    35G  0 lvm
    ├─ubuntu--vg-vm2p1    253:4    0     1M  0 part
    ├─ubuntu--vg-vm2p2    253:5    0   1.5G  0 part
    └─ubuntu--vg-vm2p3    253:6    0  33.5G  0 part

The only part that would mount is the boot part
dom0 $ sudo mount -o ro /dev/mapper/ubuntu--vg-vm2p2  /mnt/vm2/vm2p2

dom0 $ ls -alt /mnt/vm2/vm2p2/
total 101068
drwxr-xr-x 5 root root     4096 Jul 28 19:27 ..
drwxr-xr-x 4 root root     4096 Jul 27 21:18 .
-rw-r--r-- 1 root root 84797689 Jul 27 21:18 initrd.img-5.4.0-122-generic
drwxr-xr-x 4 root root     4096 Jul 27 21:15 grub
lrwxrwxrwx 1 root root       28 Jul 27 21:14 initrd.img -> initrd.img-5.4.0-122-generic
lrwxrwxrwx 1 root root       28 Jul 27 21:14 initrd.img.old -> initrd.img-5.4.0-122-generic
lrwxrwxrwx 1 root root       25 Jul 27 21:14 vmlinuz -> vmlinuz-5.4.0-122-generic
lrwxrwxrwx 1 root root       25 Jul 27 21:14 vmlinuz.old -> vmlinuz-5.4.0-122-generic
drwx------ 2 root root    16384 Jul 27 21:13 lost+found
-rw------- 1 root root 13660416 Jun 22 13:26 vmlinuz-5.4.0-122-generic                 <--- dwarf2json cannot use this
-rw-r--r-- 1 root root   237947 Jun 22 13:00 config-5.4.0-122-generic
-rw------- 1 root root  4758694 Jun 22 13:00 System.map-5.4.0-122-generic                 <--- dwarf2json uses this

dom0 $ sudo ./dwarf2json linux --system-map /mnt/vm2/vm2p2/System.map-5.4.0-122-generic --elf /mnt/vm2/vm2p2/vmlinuz-5.4.0-122-generic
Failed linux processing: could not open /mnt/vm2/vm2p2/vmlinuz-5.4.0-122-generic: bad magic number '[77 90 234 7]' in record at byte 0x0
--- this is a compressed bzimage file

dom0 $ sudo file /mnt/vm2/vm2p2/vmlinuz-5.4.0-122-generic
/mnt/vm2/vm2p2/vmlinuz-5.4.0-122-generic: Linux kernel x86 boot executable bzImage, version 5.4.0-122-generic (buildd@lcy02-amd64-095) #138-Ubuntu SMP Wed Jun 22 15:00:31 UTC 2022, RO-rootFS, swap_dev 0xD, Normal VGA


it looks like this link is not quite correct anymore WRT ubuntu: https://www.ibm.com/docs/en/linux-on-systems?topic=linuxonibm/liacf/oprofkernelsymrhel.html   *3

and this link makes me think I will need help since I cannot fined the debug linux image file:  https://github.com/volatilityfoundation/dwarf2json/issues/37   *2

since I dont see the files at all I searched for any matching files:
root@vm2:/home/cdprince# find / -name "vmlinux*"
..., etc, ...
/usr/lib/debug/boot/vmlinux-5.4.0-122-generic
..... etc, .....
 
The above file is in the debug area same as in link *3 suggests, so looking at the file:

root@vm2:/home/cdprince# ls -alt /usr/lib/debug/boot/vmlinux-5.4.0-122-generic 
-rw-r--r-- 1 root root 785425264 Jun 22 13:00 /usr/lib/debug/boot/vmlinux-5.4.0-122-generic

the file: /usr/lib/debug/boot/vmlinux-5.4.0-122-generic seems about the right size: 785MB roughly 100X the non-debug linux image file as link *2 suggests.

*****so I copied the debug boot file from:/usr/lib/debug/boot/vmlinux-5.4.0-122-generic   into   /boot/
performed kpart -a, and mounted it and then performed the 
./dwarf2json linux --system-map .... and it worked.

sudo vmi-process-list -n vm2 -j /root/vm2.json      -- Worked, but without -n and -j did not

finally, for some reason using sudo did not work, but running as root did?
$ sudo ./src/drakvuf -r /root/vm2.json -d 1
./src/drakvuf: error while loading shared libraries: libvmi.so.0: cannot open shared object file: No such file or directory
cdprince@vmmone:~/git/drakvuf$ sudo su
root@vmmone:/home/cdprince/git/drakvuf# xl list
Name                                        ID   Mem VCPUs State Time(s)
Domain-0                                     0  4096     4     r-----    1247.5
vm2                                          1  1008     2     -b----     397.0
root@vmmone:/home/cdprince/git/drakvuf# ./src/drakvuf -r /root/vm2.json -d 1
this worked!!!

--- I really hopes this helps someone ---
